cmd/otel-allocator/allocation/consistent_hashing.go

```diff
@@ -47,12 +49,14 @@ type consistentHashingAllocator struct {
 	collectors map[string]*Collector
 
 	// targetItems is a map from a target item's hash to the target items allocated state
-	targetItems map[string]*TargetItem
+	targetItems map[string]*target.Item
 
 	log logr.Logger
+
+	filterFunction prehook.Hook
 }
 
-func newConsistentHashingAllocator(log logr.Logger) Allocator {
+func newConsistentHashingAllocator(log logr.Logger, filterFunction prehook.Hook) Allocator {
```

Comment:
I would suggest using the [functional options pattern](https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis) here rather than adding new required parameters.  This will allow more flexibility in choosing how to extend the capabilities of these types.  The signature here would then become something like this:

```go
func newConsistentHashingAllocator(log logr.logger, opts ...Option) Allocator {...}
```

and it could be invoked like this:

```go
cha := newConsistentHashingAllocator(log, allocation.WithFilter(filterFn))
```

Reply:
Thanks for sharing that article, really helpful for my understanding.  I gave this implementation a try so let me know if you have any suggestions.

I added a new method to the `Allocator` interface called `SetHook` which will simply set concrete structs `filterFunction` 

Then I define

```
type AllocOption func(Allocator)

func WithFilter(hook prehook.Hook) AllocOption {
	return func(allocator Allocator) {
		allocator.SetHook(hook)
	}
}
```